File: vane-core/src/main/java/org/oddlama/vane/core/command/Param.java
package org.oddlama.vane.core.command;

import static org.oddlama.vane.util.StorageUtil.namespacedKey;

import io.papermc.paper.registry.RegistryAccess;
import io.papermc.paper.registry.RegistryKey;
import java.util.Collection;
import java.util.Collections;
import java.util.List;
import java.util.stream.Collectors;
import org.bukkit.GameMode;
import org.bukkit.OfflinePlayer;
import org.bukkit.World;
import org.bukkit.command.CommandSender;
import org.bukkit.enchantments.Enchantment;
import org.bukkit.entity.Player;
import org.bukkit.permissions.Permission;
import org.oddlama.vane.core.command.check.CheckResult;
import org.oddlama.vane.core.command.check.CombinedErrorCheckResult;
import org.oddlama.vane.core.command.check.ErrorCheckResult;
import org.oddlama.vane.core.command.check.ParseCheckResult;
import org.oddlama.vane.core.command.params.AnyParam;
import org.oddlama.vane.core.command.params.ChoiceParam;
import org.oddlama.vane.core.command.params.DynamicChoiceParam;
import org.oddlama.vane.core.command.params.FixedParam;
import org.oddlama.vane.core.command.params.SentinelExecutorParam;
import org.oddlama.vane.core.functional.Consumer1;
import org.oddlama.vane.core.functional.Consumer2;
import org.oddlama.vane.core.functional.Consumer3;
import org.oddlama.vane.core.functional.Consumer4;
import org.oddlama.vane.core.functional.Consumer5;
import org.oddlama.vane.core.functional.Consumer6;
import org.oddlama.vane.core.functional.Function1;
import org.oddlama.vane.core.functional.Function2;
import org.oddlama.vane.core.functional.Function3;
import org.oddlama.vane.core.functional.Function4;
import org.oddlama.vane.core.functional.Function5;
import org.oddlama.vane.core.functional.Function6;
import org.oddlama.vane.core.module.Module;

@SuppressWarnings("overloads")
public interface Param {
    public List<Param> getParams();

    public default boolean requirePlayer(CommandSender sender) {
        if (!(sender instanceof Player)) {
            getCommand().getModule().core.langCommandNotAPlayer.send(sender);
            return false;
        }

        return true;
    }

    public default boolean isExecutor() {
        return false;
    }

    public default void addParam(Param param) {
        if (param.isExecutor() && getParams().stream().anyMatch(p -> p.isExecutor())) {
            throw new RuntimeException("Cannot define multiple executors for the same parameter! This is a bug.");
        }
        getParams().add(param);
    }

    public default <T1> void execPlayer(Consumer1<T1> f) {
        addParam(new SentinelExecutorParam<>(getCommand(), f, this::requirePlayer, i -> i == 0));
    }

    public default <T1, T2> void execPlayer(Consumer2<T1, T2> f) {
        addParam(new SentinelExecutorParam<>(getCommand(), f, this::requirePlayer, i -> i == 0));
    }

    public default <T1, T2, T3> void execPlayer(Consumer3<T1, T2, T3> f) {
        addParam(new SentinelExecutorParam<>(getCommand(), f, this::requirePlayer, i -> i == 0));
    }

    public default <T1, T2, T3, T4> void execPlayer(Consumer4<T1, T2, T3, T4> f) {
        addParam(new SentinelExecutorParam<>(getCommand(), f, this::requirePlayer, i -> i == 0));
    }

    public default <T1, T2, T3, T4, T5> void execPlayer(Consumer5<T1, T2, T3, T4, T5> f) {
        addParam(new SentinelExecutorParam<>(getCommand(), f, this::requirePlayer, i -> i == 0));
    }

    public default <T1, T2, T3, T4, T5, T6> void execPlayer(Consumer6<T1, T2, T3, T4, T5, T6> f) {
        addParam(new SentinelExecutorParam<>(getCommand(), f, this::requirePlayer, i -> i == 0));
    }

    public default <T1> void execPlayer(Function1<T1, Boolean> f) {
        addParam(new SentinelExecutorParam<>(getCommand(), f, this::requirePlayer, i -> i == 0));
    }

    public default <T1, T2> void execPlayer(Function2<T1, T2, Boolean> f) {
        addParam(new SentinelExecutorParam<>(getCommand(), f, this::requirePlayer, i -> i == 0));
    }

    public default <T1, T2, T3> void execPlayer(Function3<T1, T2, T3, Boolean> f) {
        addParam(new SentinelExecutorParam<>(getCommand(), f, this::requirePlayer, i -> i == 0));
    }

    public default <T1, T2, T3, T4> void execPlayer(Function4<T1, T2, T3, T4, Boolean> f) {
        addParam(new SentinelExecutorParam<>(getCommand(), f, this::requirePlayer, i -> i == 0));
    }

    public default <T1, T2, T3, T4, T5> void execPlayer(Function5<T1, T2, T3, T4, T5, Boolean> f) {
        addParam(new SentinelExecutorParam<>(getCommand(), f, this::requirePlayer, i -> i == 0));
    }

    public default <T1, T2, T3, T4, T5, T6> void execPlayer(Function6<T1, T2, T3, T4, T5, T6, Boolean> f) {
        addParam(new SentinelExecutorParam<>(getCommand(), f, this::requirePlayer, i -> i == 0));
    }

    public default <T1> void exec(Consumer1<T1> f) {
        addParam(new SentinelExecutorParam<>(getCommand(), f));
    }

    public default <T1, T2> void exec(Consumer2<T1, T2> f) {
        addParam(new SentinelExecutorParam<>(getCommand(), f));
    }

    public default <T1, T2, T3> void exec(Consumer3<T1, T2, T3> f) {
        addParam(new SentinelExecutorParam<>(getCommand(), f));
    }

    public default <T1, T2, T3, T4> void exec(Consumer4<T1, T2, T3, T4> f) {
        addParam(new SentinelExecutorParam<>(getCommand(), f));
    }

    public default <T1, T2, T3, T4, T5> void exec(Consumer5<T1, T2, T3, T4, T5> f) {
        addParam(new SentinelExecutorParam<>(getCommand(), f));
    }

    public default <T1, T2, T3, T4, T5, T6> void exec(Consumer6<T1, T2, T3, T4, T5, T6> f) {
        addParam(new SentinelExecutorParam<>(getCommand(), f));
    }

    public default <T1> void exec(Function1<T1, Boolean> f) {
        addParam(new SentinelExecutorParam<>(getCommand(), f));
    }

    public default <T1, T2> void exec(Function2<T1, T2, Boolean> f) {
        addParam(new SentinelExecutorParam<>(getCommand(), f));
    }

    public default <T1, T2, T3> void exec(Function3<T1, T2, T3, Boolean> f) {
        addParam(new SentinelExecutorParam<>(getCommand(), f));
    }

    public default <T1, T2, T3, T4> void exec(Function4<T1, T2, T3, T4, Boolean> f) {
        addParam(new SentinelExecutorParam<>(getCommand(), f));
    }

    public default <T1, T2, T3, T4, T5> void exec(Function5<T1, T2, T3, T4, T5, Boolean> f) {
        addParam(new SentinelExecutorParam<>(getCommand(), f));
    }

    public default <T1, T2, T3, T4, T5, T6> void exec(Function6<T1, T2, T3, T4, T5, T6, Boolean> f) {
        addParam(new SentinelExecutorParam<>(getCommand(), f));
    }

    public default Param anyString() {
        return any("string", str -> str);
    }

    public default <T> AnyParam<? extends T> any(String argumentType, Function1<String, ? extends T> fromString) {
        var p = new AnyParam<>(getCommand(), argumentType, fromString);
        addParam(p);
        return p;
    }

    public default FixedParam<String> fixed(String fixed) {
        return fixed(fixed, str -> str);
    }

    public default <T> FixedParam<T> fixed(T fixed, Function1<T, String> toString) {
        var p = new FixedParam<>(getCommand(), fixed, toString);
        addParam(p);
        return p;
    }

    public default Param choice(Collection<String> choices) {
        return choice("choice", choices, str -> str);
    }

    public default <T> ChoiceParam<T> choice(
        String argumentType,
        Collection<? extends T> choices,
        Function1<T, String> toString
    ) {
        var p = new ChoiceParam<>(getCommand(), argumentType, choices, toString);
        addParam(p);
        return p;
    }

    public default <T> DynamicChoiceParam<T> choice(
        String argumentType,
        Function1<CommandSender, Collection<? extends T>> choices,
        Function2<CommandSender, T, String> toString,
        Function2<CommandSender, String, ? extends T> fromString
    ) {
        var p = new DynamicChoiceParam<>(getCommand(), argumentType, choices, toString, fromString);
        addParam(p);
        return p;
    }

    public default DynamicChoiceParam<Module<?>> chooseModule() {
        return choice(
            "module",
            sender -> getCommand().getModule().core.getModules(),
            (sender, m) -> m.getAnnotationName(),
            (sender, str) ->
                getCommand()
                    .getModule()
                    .core.getModules()
                    .stream()
                    .filter(k -> k.getAnnotationName().equalsIgnoreCase(str))
                    .findFirst()
                    .orElse(null)
        );
    }

    public default DynamicChoiceParam<World> chooseWorld() {
        return choice(
            "world",
            sender -> getCommand().getModule().getServer().getWorlds(),
            (sender, w) -> w.getName().toLowerCase(),
            (sender, str) ->
                getCommand()
                    .getModule()
                    .getServer()
                    .getWorlds()
                    .stream()
                    .filter(w -> w.getName().equalsIgnoreCase(str))
                    .findFirst()
                    .orElse(null)
        );
    }

    public default DynamicChoiceParam<OfflinePlayer> chooseAnyPlayer() {
        return choice(
            "any_player",
            sender -> getCommand().getModule().getOfflinePlayersWithValidName(),
            (sender, p) -> p.getName(),
            (sender, str) ->
                getCommand()
                    .getModule()
                    .getOfflinePlayersWithValidName()
                    .stream()
                    .filter(k -> k.getName().equalsIgnoreCase(str))
                    .findFirst()
                    .orElse(null)
        );
    }

    public default DynamicChoiceParam<Player> chooseOnlinePlayer() {
        return choice(
            "online_player",
            sender -> getCommand().getModule().getServer().getOnlinePlayers(),
            (sender, p) -> p.getName(),
            (sender, str) ->
                getCommand()
                    .getModule()
                    .getServer()
                    .getOnlinePlayers()
                    .stream()
                    .filter(k -> k.getName().equalsIgnoreCase(str))
                    .findFirst()
                    .orElse(null)
        );
    }

    // TODO (minor): Make choosePermission filter results based on the previously
    // specified player.
    public default DynamicChoiceParam<Permission> choosePermission() {
        return choice(
            "permission",
            sender -> getCommand().getModule().getServer().getPluginManager().getPermissions(),
            (sender, p) -> p.getName(),
            (sender, str) -> getCommand().getModule().getServer().getPluginManager().getPermission(str)
        );
    }

    public default ChoiceParam<GameMode> chooseGamemode() {
        return choice("gamemode", List.of(GameMode.values()), m -> m.name().toLowerCase()).ignoreCase();
    }

    public default DynamicChoiceParam<Enchantment> chooseEnchantment() {
        return chooseEnchantment((sender, e) -> true);
    }

    public default DynamicChoiceParam<Enchantment> chooseEnchantment(
        final Function2<CommandSender, Enchantment, Boolean> filter
    ) {
        return choice(
            "enchantment",
            sender ->
                RegistryAccess.registryAccess()
                    .getRegistry(RegistryKey.ENCHANTMENT)
                    .stream()
                    .filter(e -> filter.apply(sender, e))
                    .collect(Collectors.toList()),
            (sender, e) -> e.getKey().toString(),
            (sender, str) -> {
                var parts = str.split(":");
                if (parts.length != 2) {
                    return null;
                }
                var e = RegistryAccess.registryAccess()
                    .getRegistry(RegistryKey.ENCHANTMENT)
                    .get(namespacedKey(parts[0], parts[1]));
                if (!filter.apply(sender, e)) {
                    return null;
                }
                return e;
            }
        );
    }

    public default CheckResult checkAcceptDelegate(CommandSender sender, String[] args, int offset) {
        if (getParams().isEmpty()) {
            throw new RuntimeException("Encountered parameter without sentinel! This is a bug.");
        }

        // Delegate to children
        var results = getParams().stream().map(p -> p.checkAccept(sender, args, offset + 1)).toList();

        // Return the first executor result, if any
        for (var r : results) {
            if (r.good()) {
                return r;
            }
        }

        // Only retain errors from maximum depth
        var maxDepth = results.stream().map(r -> r.depth()).reduce(0, Integer::max);

        var errors = results
            .stream()
            .filter(r -> r.depth() == maxDepth)
            .map(ErrorCheckResult.class::cast)
            .collect(Collectors.toList());

        // If there is only a single max-depth sub-error, propagate it.
        // Otherwise, combine multiple errors into new error.
        if (errors.size() == 1) {
            return errors.get(0);
        } else {
            return new CombinedErrorCheckResult(errors);
        }
    }

    public default CheckResult checkAccept(CommandSender sender, String[] args, int offset) {
        var result = checkParse(sender, args, offset);
        if (!(result instanceof ParseCheckResult)) {
            return result;
        }

        var p = (ParseCheckResult) result;
        return checkAcceptDelegate(sender, args, offset).prepend(p.argumentType(), p.parsed(), p.includeParam());
    }

    public default List<String> buildCompletionsDelegate(CommandSender sender, String[] args, int offset) {
        if (getParams().isEmpty()) {
            throw new RuntimeException("Encountered parameter without sentinel! This is a bug.");
        }

        // Delegate to children
        return getParams()
            .stream()
            .map(p -> p.buildCompletions(sender, args, offset + 1))
            .flatMap(Collection::stream)
            .collect(Collectors.toList());
    }

    public default List<String> buildCompletions(CommandSender sender, String[] args, int offset) {
        if (offset < args.length - 1) {
            // We are not the last argument.
            // Delegate completion to children if the param accepts the given arguments,
            // return no completions if it doesn't
            if (checkParse(sender, args, offset) instanceof ParseCheckResult) {
                return buildCompletionsDelegate(sender, args, offset);
            } else {
                return Collections.emptyList();
            }
        } else {
            // We are the parameter that needs to be completed.
            // Offer (partial) completions if our depth is the completion depth
            return completionsFor(sender, args, offset);
        }
    }

    public List<String> completionsFor(CommandSender sender, String[] args, int offset);

    public CheckResult checkParse(CommandSender sender, String[] args, int offset);

    public Command<?> getCommand();
}

File: vane-core/src/main/kotlin/org/oddlama/vane/core/command/Param.kt
package org.oddlama.vane.core.command

import io.papermc.paper.registry.RegistryAccess
import io.papermc.paper.registry.RegistryKey
import org.bukkit.GameMode
import org.bukkit.OfflinePlayer
import org.bukkit.World
import org.bukkit.command.CommandSender
import org.bukkit.enchantments.Enchantment
import org.bukkit.entity.Player
import org.bukkit.permissions.Permission
import org.oddlama.vane.core.command.check.CheckResult
import org.oddlama.vane.core.command.check.CombinedErrorCheckResult
import org.oddlama.vane.core.command.check.ErrorCheckResult
import org.oddlama.vane.core.command.check.ParseCheckResult
import org.oddlama.vane.core.command.params.*
import org.oddlama.vane.core.functional.*
import org.oddlama.vane.core.module.Module
import org.oddlama.vane.util.StorageUtil.namespacedKey

interface Param {
    val params: MutableList<Param?>

    fun requirePlayer(sender: CommandSender?): Boolean {
        if (sender !is Player) {
            this.command!!.module!!.core?.langCommandNotAPlayer?.send(sender)
            return false
        }

        return true
    }

    val isExecutor: Boolean
        get() = false

    fun addParam(param: Param) {
        if (param.isExecutor && params.any { it?.isExecutor == true }) {
            throw RuntimeException("Cannot define multiple executors for the same parameter! This is a bug.")
        }
        params.add(param)
    }

    fun <T1> execPlayer(f: Consumer1<T1?>?) {
        addParam(SentinelExecutorParam<Consumer1<T1?>?>(command, f, { requirePlayer(it) }, { it == 0 }))
    }

    fun <T1, T2> execPlayer(f: Consumer2<T1?, T2?>?) {
        addParam(SentinelExecutorParam<Consumer2<T1?, T2?>?>(command, f, { requirePlayer(it) }, { it == 0 }))
    }

    fun <T1, T2, T3> execPlayer(f: Consumer3<T1?, T2?, T3?>?) {
        addParam(SentinelExecutorParam<Consumer3<T1?, T2?, T3?>?>(command, f, { requirePlayer(it) }, { it == 0 }))
    }

    fun <T1, T2, T3, T4> execPlayer(f: Consumer4<T1?, T2?, T3?, T4?>?) {
        addParam(SentinelExecutorParam<Consumer4<T1?, T2?, T3?, T4?>?>(command, f, { requirePlayer(it) }, { it == 0 }))
    }

    fun <T1, T2, T3, T4, T5> execPlayer(f: Consumer5<T1?, T2?, T3?, T4?, T5?>?) {
        addParam(SentinelExecutorParam<Consumer5<T1?, T2?, T3?, T4?, T5?>?>(command, f, { requirePlayer(it) }, { it == 0 }))
    }

    fun <T1, T2, T3, T4, T5, T6> execPlayer(f: Consumer6<T1?, T2?, T3?, T4?, T5?, T6?>?) {
        addParam(SentinelExecutorParam<Consumer6<T1?, T2?, T3?, T4?, T5?, T6?>?>(command, f, { requirePlayer(it) }, { it == 0 }))
    }

    fun <T1> execPlayer(f: Function1<T1?, Boolean?>?) {
        addParam(SentinelExecutorParam<Function1<T1?, Boolean?>?>(command, f, Function1 { requirePlayer(it) }, Function1 { it == 0 }))
    }

    fun <T1, T2> execPlayer(f: Function2<T1?, T2?, Boolean?>?) {
        addParam(SentinelExecutorParam<Function2<T1?, T2?, Boolean?>?>(command, f, Function1 { requirePlayer(it) }, Function1 { it == 0 }))
    }

    fun <T1, T2, T3> execPlayer(f: Function3<T1?, T2?, T3?, Boolean?>?) {
        addParam(SentinelExecutorParam<Function3<T1?, T2?, T3?, Boolean?>?>(command, f, Function1 { requirePlayer(it) }, Function1 { it == 0 }))
    }

    fun <T1, T2, T3, T4> execPlayer(f: Function4<T1?, T2?, T3?, T4?, Boolean?>?) {
        addParam(SentinelExecutorParam<Function4<T1?, T2?, T3?, T4?, Boolean?>?>(command, f, Function1 { requirePlayer(it) }, Function1 { it == 0 }))
    }

    fun <T1, T2, T3, T4, T5> execPlayer(f: Function5<T1?, T2?, T3?, T4?, T5?, Boolean?>?) {
        addParam(SentinelExecutorParam<Function5<T1?, T2?, T3?, T4?, T5?, Boolean?>?>(command, f, Function1 { requirePlayer(it) }, Function1 { it == 0 }))
    }

    fun <T1, T2, T3, T4, T5, T6> execPlayer(f: Function6<T1?, T2?, T3?, T4?, T5?, T6?, Boolean?>?) {
        addParam(SentinelExecutorParam<Function6<T1?, T2?, T3?, T4?, T5?, T6?, Boolean?>?>(command, f, Function1 { requirePlayer(it) }, Function1 { it == 0 }))
    }

    fun <T1> exec(f: Consumer1<T1?>?) {
        addParam(SentinelExecutorParam<Consumer1<T1?>?>(command, f))
    }

    fun <T1, T2> exec(f: Consumer2<T1?, T2?>?) {
        addParam(SentinelExecutorParam<Consumer2<T1?, T2?>?>(command, f))
    }

    fun <T1, T2, T3> exec(f: Consumer3<T1?, T2?, T3?>?) {
        addParam(SentinelExecutorParam<Consumer3<T1?, T2?, T3?>?>(command, f))
    }

    fun <T1, T2, T3, T4> exec(f: Consumer4<T1?, T2?, T3?, T4?>?) {
        addParam(SentinelExecutorParam<Consumer4<T1?, T2?, T3?, T4?>?>(command, f))
    }

    fun <T1, T2, T3, T4, T5> exec(f: Consumer5<T1?, T2?, T3?, T4?, T5?>?) {
        addParam(SentinelExecutorParam<Consumer5<T1?, T2?, T3?, T4?, T5?>?>(command, f))
    }

    fun <T1, T2, T3, T4, T5, T6> exec(f: Consumer6<T1?, T2?, T3?, T4?, T5?, T6?>?) {
        addParam(SentinelExecutorParam<Consumer6<T1?, T2?, T3?, T4?, T5?, T6?>?>(command, f))
    }

    fun <T1> exec(f: Function1<T1?, Boolean?>?) {
        addParam(SentinelExecutorParam<Function1<T1?, Boolean?>?>(command, f))
    }

    fun <T1, T2> exec(f: Function2<T1?, T2?, Boolean?>?) {
        addParam(SentinelExecutorParam<Function2<T1?, T2?, Boolean?>?>(command, f))
    }

    fun <T1, T2, T3> exec(f: Function3<T1?, T2?, T3?, Boolean?>?) {
        addParam(SentinelExecutorParam<Function3<T1?, T2?, T3?, Boolean?>?>(command, f))
    }

    fun <T1, T2, T3, T4> exec(f: Function4<T1?, T2?, T3?, T4?, Boolean?>?) {
        addParam(SentinelExecutorParam<Function4<T1?, T2?, T3?, T4?, Boolean?>?>(command, f))
    }

    fun <T1, T2, T3, T4, T5> exec(f: Function5<T1?, T2?, T3?, T4?, T5?, Boolean?>?) {
        addParam(SentinelExecutorParam<Function5<T1?, T2?, T3?, T4?, T5?, Boolean?>?>(command, f))
    }

    fun <T1, T2, T3, T4, T5, T6> exec(f: Function6<T1?, T2?, T3?, T4?, T5?, T6?, Boolean?>?) {
        addParam(SentinelExecutorParam<Function6<T1?, T2?, T3?, T4?, T5?, T6?, Boolean?>?>(command, f))
    }

    fun anyString(): Param = any<String?>("string") { str -> str }

    fun <T> any(argumentType: String, fromString: Function1<String?, out T?>): AnyParam<out T?> {
        val p = AnyParam<T?>(this.command, argumentType, fromString)
        addParam(p)
        return p
    }

    fun fixed(fixed: String?): FixedParam<String?> = fixed<String?>(fixed) { str -> str }

    fun <T> fixed(fixed: T?, toString: Function1<T?, String?>): FixedParam<T?> {
        val p = FixedParam<T?>(this.command, fixed, toString)
        addParam(p)
        return p
    }

    fun choice(choices: Collection<String?>): Param = choice<String?>("choice", choices) { str -> str }

    fun <T> choice(
        argumentType: String,
        choices: Collection<T?>,
        toString: Function1<T?, String?>
    ): ChoiceParam<T?> {
        val p = ChoiceParam<T?>(this.command, argumentType, choices, toString)
        addParam(p)
        return p
    }

    fun <T> choice(
        argumentType: String,
        choices: Function1<CommandSender?, Collection<T?>?>,
        toString: Function2<CommandSender?, T?, String?>,
        fromString: Function2<CommandSender?, String?, out T?>
    ): DynamicChoiceParam<T?> {
        val p = DynamicChoiceParam<T?>(this.command, argumentType, choices, toString, fromString)
        addParam(p)
        return p
    }

    fun chooseModule(): DynamicChoiceParam<Module<*>?> {
        return choice<Module<*>?>(
            "module",
            Function1 { _ -> this.command!!.module!!.core?.modules ?: mutableSetOf<Module<*>>() },
            Function2 { _, m -> m!!.annotationName },
            Function2 { _, str ->
                val core = this.command!!.module!!.core ?: return@Function2 null
                core.modules.firstOrNull { it?.annotationName.equals(str, ignoreCase = true) }
            }
        )
    }

    fun chooseWorld(): DynamicChoiceParam<World?> {
        return choice<World?>(
            "world",
            { _ -> this.command!!.module!!.server.worlds },
            { _, w -> w!!.name.lowercase() },
            Function2 { _, str ->
                this.command!!.module!!.server.worlds
                    .firstOrNull { it.name.equals(str, ignoreCase = true) }
            }
        )
    }

    fun chooseAnyPlayer(): DynamicChoiceParam<OfflinePlayer?> {
        return choice<OfflinePlayer?>(
            "any_player",
            { _ -> this.command!!.module!!.offlinePlayersWithValidName.toMutableList() },
            { _, p -> p!!.name },
            Function2 { _, str ->
                this.command!!.module!!.offlinePlayersWithValidName
                    .firstOrNull { it.name.equals(str, ignoreCase = true) }
            }
        )
    }

    fun chooseOnlinePlayer(): DynamicChoiceParam<Player?> {
        return choice<Player?>(
            "online_player",
            { _ -> this.command!!.module!!.server.onlinePlayers.toMutableList() },
            { _, p -> p!!.name },
            Function2 { _, str ->
                this.command!!.module!!.server.onlinePlayers
                    .firstOrNull { it.name.equals(str, ignoreCase = true) }
            }
        )
    }

    // TODO (minor): Make choosePermission filter results based on the previously
    // specified player.
    fun choosePermission(): DynamicChoiceParam<Permission?> {
        return choice<Permission?>(
            "permission",
            { _ -> this.command!!.module!!.server.pluginManager.permissions },
            { _, p -> p!!.name },
            Function2 { _, str -> this.command!!.module!!.server.pluginManager.getPermission(str!!) }
        )
    }

    fun chooseGamemode(): ChoiceParam<GameMode?> =
        choice<GameMode?>("gamemode", GameMode.entries) { m -> m!!.name.lowercase() }.ignoreCase()

    fun chooseEnchantment(
        filter: Function2<CommandSender?, Enchantment?, Boolean?> = Function2 { _, _ -> true }
    ): DynamicChoiceParam<Enchantment?> {
        return choice<Enchantment?>(
            "enchantment",
            { _ ->
                RegistryAccess.registryAccess()
                    .getRegistry(RegistryKey.ENCHANTMENT)
                    .filter { filter.apply(null, it) == true }
                    .toMutableList()
            },
            { _, e -> e!!.key.toString() },
            Function2 { _, str ->
                val parts = str?.split(":") ?: return@Function2 null
                if (parts.size != 2) return@Function2 null
                val e = RegistryAccess.registryAccess()
                    .getRegistry(RegistryKey.ENCHANTMENT)
                    .get(namespacedKey(parts[0], parts[1]))
                e.takeIf { filter.apply(null, it) == true }
            }
        )
    }

    fun checkAcceptDelegate(sender: CommandSender?, args: Array<String?>?, offset: Int): CheckResult? {
        require(params.isNotEmpty()) { "Encountered parameter without sentinel! This is a bug." }

        val results = params.mapNotNull { it!!.checkAccept(sender, args, offset + 1) }

        // Return the first executor result, if any
        results.firstOrNull { it.good() }?.let { return it }

        // Only retain errors from maximum depth
        val maxDepth = results.maxOf { it.depth() }
        val errors = results
            .filter { it.depth() == maxDepth }
            .filterIsInstance<ErrorCheckResult>()

        // If there is only a single max-depth sub-error, propagate it.
        // Otherwise, combine multiple errors into new error.
        return if (errors.size == 1) errors[0] else CombinedErrorCheckResult(errors)
    }

    fun checkAccept(sender: CommandSender?, args: Array<String?>?, offset: Int): CheckResult? {
        val result = checkParse(sender, args, offset)
        if (result !is ParseCheckResult) {
            return result
        }

        return checkAcceptDelegate(sender, args, offset)!!.prepend(
            result.argumentType,
            result.parsed, result.includeParam)
    }

    fun buildCompletionsDelegate(sender: CommandSender?, args: Array<String?>, offset: Int): MutableList<String?> {
        require(params.isNotEmpty()) { "Encountered parameter without sentinel! This is a bug." }

        return params
            .flatMap { it!!.buildCompletions(sender, args, offset + 1) ?: emptyList() }
            .toMutableList()
    }

    fun buildCompletions(sender: CommandSender?, args: Array<String?>, offset: Int): MutableList<String?>? {
        return if (offset < args.size - 1) {
            // We are not the last argument.
            // Delegate completion to children if the param accepts the given arguments,
            // return no completions if it doesn't
            if (checkParse(sender, args, offset) is ParseCheckResult) {
                buildCompletionsDelegate(sender, args, offset)
            } else {
                mutableListOf()
            }
        } else {
            // We are the parameter that needs to be completed.
            // Offer (partial) completions if our depth is the completion depth
            completionsFor(sender, args, offset)
        }
    }

    fun completionsFor(sender: CommandSender?, args: Array<String?>?, offset: Int): MutableList<String?>?

    fun checkParse(sender: CommandSender?, args: Array<String?>?, offset: Int): CheckResult?

    val command: Command<*>?
}
